File: DataStructure_Algorithm/Algorithm/Python/Graphic/Dijkstra/Dijkstra.py
#coding:utf-8
import sys

MAX_VALUE = 999999999
vertexNum = 5

d = [0,0,0,0,0]
pre = [0,0,0,0,0]
visited = [False,False,False,False,False]
adjacentMatrix = [
	[ 0,10, 0, 5, 0],
	[ 0, 0, 1, 2, 0],
	[ 0, 0, 0, 0, 4],
	[ 0, 3, 9, 0, 2],
	[ 7, 0, 6, 0, 0]
]

def InitializeSingleSource(s):
	for u in range(vertexNum):
		d[u] = MAX_VALUE
		pre[u] = None
	d[s] = 0

def ExtractMin():
	distance = MAX_VALUE
	m = -1
	for u in range(vertexNum):
		if visited[u] == False and d[u] < distance:
			distance = d[u]
			m = u
	return m
	
def Relax(u,v,w):
	if d[v] > d[u] + w:
		d[v] = d[u] + w
		pre[v] = u
		
def Dijkstra(s):
	InitializeSingleSource(s)
	print d
	for i in range(vertexNum):
		u = ExtractMin()
		visited[u] = True
		for v in range(vertexNum):
			if adjacentMatrix[u][v] > 0 and visited[v] == False:
				Relax(u,v,adjacentMatrix[u][v])
				
def PrintPath(s):
	if pre[s] == None:
		sys.stdout.write('%d'%s)
	else:
		PrintPath(pre[s])
		sys.stdout.write('-->%d'%s)
		
Dijkstra(1)
print d
print visited
print pre
PrintPath(0)
print ' '
PrintPath(2)
print ' '
PrintPath(3)
print ' '
PrintPath(4)
print ' '
				

File: DataStructure_Algorithm/Algorithm/C_C++/Graphic/Dijkstra/Dijkstra.cpp
#include <iostream>
#include <algorithm>

using namespace std;

const int LEN = 20;
const int MAX_VALUE = 999999999;

int ExtractMin(int n, int *visited, int *distance)
{
	int dis = MAX_VALUE;
	int minIdx = -1;
	for (int i = 0; i < n; i++)
	{
		if (visited[i] == 0 && distance[i] < dis)
		{
			dis = distance[i];
			minIdx = i;
		}
	}

	return minIdx;
}

void Relax(int u, int v, int dis, int *distance, int *parent)
{
	if (distance[v] > distance[u] + dis)
	{
		distance[v] = distance[u] + dis;
		parent[v] = u;
	}
}

void PrintPath(int vertex, int *parent)
{
	if (parent[vertex] == -1)
	{
		cout << vertex;
		return;
	}
	PrintPath(parent[vertex], parent);
	cout << "->" << vertex;
}

void Dijkstra(int *adjacentMatrix, int startVertex, int n)
{
	// initial
	int visited[10];
	int parent[10];
	int distance[10];
	fill(visited, visited + n, 0);
	fill(parent, parent + n, -1);
	fill(distance, distance + n, MAX_VALUE);

	distance[startVertex] = 0;
	for (int i = 0; i < n; i++)
	{
		int vertex = ExtractMin(n, visited, distance);
		if (vertex == -1)
			continue;
		visited[vertex] = 1;
		for (int j = 0; j < n; j++)
		{
			if (adjacentMatrix[vertex*n + j] > 0 && visited[j] == 0)
				Relax(vertex, j, adjacentMatrix[vertex*n + j], distance, parent);
		}
	}

	for (int i = 0; i < n; i++)
	{
		PrintPath(i, parent);
		cout << endl;
	}
}

int main()
{
	int adjacentMatrix[100];

	FILE *stream;
	freopen_s(&stream, "Dijkstra.txt", "r", stdin);
	freopen_s(&stream, "out.txt", "w", stdout);
	int n;
	cin >> n;
	for (int i = 0; i < n*n; i++)
		cin >> adjacentMatrix[i];

	Dijkstra(adjacentMatrix, 0, n);

	fclose(stdin);
	fclose(stdout);
	//system("pause");

	return 0;
}
